python/tank/platform/engine.py

```diff
@@ -959,33 +1001,59 @@ def __create_main_thread_invoker(self):
         """
         Create the object used to invoke function calls on the main thread when
         called from a different thread.
-        
+
         :returns:  Invoker instance
         """
         if self.has_ui:
             from .qt import QtGui, QtCore
             if QtGui and QtCore:
                 class Invoker(QtCore.QObject):
+                    """
+                    Invoker class - implements a mechanism to execute a function with arbitrary
+                    args in the main thread.
+                    """
                     def __init__(self):
+                        """
+                        Construction
+                        """
                         QtCore.QObject.__init__(self)
+                        self._lock = threading.Lock()
+                        self._fn = None
                         self._res = None
-                        
+
                     def invoke(self, fn, *args, **kwargs):
-                        self._fn = lambda: fn(*args, **kwargs) 
-                        self._res = None
-                        
-                        QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
-                        
-                        return self._res
-                
+                        """
+                        Invoke the specified function with the specified args in the main thread
+
+                        :param fn:          The function to execute in the main thread
+                        :param *args:       Args for the function
+                        :param **kwargs:    Named arguments for the function
+                        :returns:           The result returned by the function
+                        """
+                        # acquire lock to ensure that the function and result are not overwritten 
+                        # by syncrounous calls to this method from different threads
+                        self._lock.acquire()
```

Comment:
ok, I had to go and do some tests for this one!  To be honest, I think the original code for this came off a stack-overflow post so no harm in checking ;)

Turns out the cost of locking via the mutex is marginally less than the cost of creating and moving an invoker instance to the main thread per-call (which is what it would need to allow for correct garbage collection) and the current method is also a bit more stable in old PySide versions/harder to break (something I'm becoming a bit of an expert on!).

As this has been tested in the wild, I'm inclined not to change it unless you think it would be a cleaner implementation?  One thing it could potentially do would be to create the invoker when first needed instead of as part of construction - that would be a good change to make I think..?